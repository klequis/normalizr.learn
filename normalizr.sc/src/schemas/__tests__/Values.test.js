/* eslint-env jest */
import { denormalize, normalize, schema } from '../../';

describe(`${schema.Values.name} normalization`, () => {
  it('normalizes the values of an object with the given schema', () => {
    const cat = new schema.Entity('cats');
    const dog = new schema.Entity('dogs');
    const valuesSchema = new schema.Values({
      dogs: dog,
      cats: cat
    }, (entity, key) => entity.type);

    expect(normalize({
      fido: { id: 1, type: 'dogs' },
      fluffy: { id: 1, type: 'cats' }
    }, valuesSchema)).toMatchSnapshot();
  });

  it('can use a function to determine the schema when normalizing', () => {
    const cat = new schema.Entity('cats');
    const dog = new schema.Entity('dogs');
    const valuesSchema = new schema.Values({
      dogs: dog,
      cats: cat
    }, (entity, key) => `${entity.type}s`);

    expect(normalize({
      fido: { id: 1, type: 'dog' },
      fluffy: { id: 1, type: 'cat' },
      jim: { id: 2, type: 'lizard' }
    }, valuesSchema)).toMatchSnapshot();
  });

  it('filters out null and undefined values', () => {
    const cat = new schema.Entity('cats');
    const dog = new schema.Entity('dogs');
    const valuesSchema = new schema.Values({
      dogs: dog,
      cats: cat
    }, (entity, key) => entity.type);

    expect(normalize({
      fido: undefined,
      milo: null,
      fluffy: { id: 1, type: 'cats' }
    }, valuesSchema)).toMatchSnapshot();
  });
});

describe(`${schema.Values.name} denormalization`, () => {
  it('denormalizes the values of an object with the given schema', () => {
    const cat = new schema.Entity('cats');
    const dog = new schema.Entity('dogs');
    const valuesSchema = new schema.Values({
      dogs: dog,
      cats: cat
    }, (entity, key) => entity.type);

    const entities = {
      cats: { 1: { id: 1, type: 'cats' } },
      dogs: { 1: { id: 1, type: 'dogs' } }
    };

    expect(denormalize({
      fido: { id: 1, schema: 'dogs' },
      fluffy: { id: 1, schema: 'cats' }
    }, valuesSchema, entities)).toMatchSnapshot();
  });
});
