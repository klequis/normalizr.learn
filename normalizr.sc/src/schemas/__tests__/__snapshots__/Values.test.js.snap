exports[`ValuesSchema denormalization denormalizes the values of an object with the given schema 1`] = `
Object {
  "fido": Object {
    "id": 1,
    "type": "dogs",
  },
  "fluffy": Object {
    "id": 1,
    "type": "cats",
  },
}
`;

exports[`ValuesSchema normalization can use a function to determine the schema when normalizing 1`] = `
Object {
  "entities": Object {
    "cats": Object {
      "1": Object {
        "id": 1,
        "type": "cat",
      },
    },
    "dogs": Object {
      "1": Object {
        "id": 1,
        "type": "dog",
      },
    },
  },
  "result": Object {
    "fido": Object {
      "id": 1,
      "schema": "dogs",
    },
    "fluffy": Object {
      "id": 1,
      "schema": "cats",
    },
    "jim": Object {
      "id": 2,
      "type": "lizard",
    },
  },
}
`;

exports[`ValuesSchema normalization filters out null and undefined values 1`] = `
Object {
  "entities": Object {
    "cats": Object {
      "1": Object {
        "id": 1,
        "type": "cats",
      },
    },
  },
  "result": Object {
    "fluffy": Object {
      "id": 1,
      "schema": "cats",
    },
  },
}
`;

exports[`ValuesSchema normalization normalizes the values of an object with the given schema 1`] = `
Object {
  "entities": Object {
    "cats": Object {
      "1": Object {
        "id": 1,
        "type": "cats",
      },
    },
    "dogs": Object {
      "1": Object {
        "id": 1,
        "type": "dogs",
      },
    },
  },
  "result": Object {
    "fido": Object {
      "id": 1,
      "schema": "dogs",
    },
    "fluffy": Object {
      "id": 1,
      "schema": "cats",
    },
  },
}
`;
